### PR TITLE
Add support for relative target/exchange/nameserver values

### PR DESCRIPTION
**Pull request by ross**
This adds support and tests for relative values as target (CNAME, NS, ...), exchanges (MX), etc. 

They are validations that will object by default and can be turned into warnings with `lenient`. This should be completely backwards compatible.

/cc Fixes https://github.com/octodns/octodns/issues/987 @rafaelfuchsbr

### CODE REVIEW

**Comment by ross** [comment]
Main question here is whether we should encourage fqdn with a validation warning/error or just allow them outright. The PR currently landed on encourage FQDNs as a best practice. 

After testing with some providers though so far they're all appending `.`'s onto the end of the relative names when queried via DNS:

```console
(env) coho:octodns ross$ dig +short relative.xormedia.com
www.
```

Things do come back from the API without the `.`, but looks like at least some of the providers append the `.` themselves and always end up with changes:

```
********************************************************************************
* exxampled.com.
********************************************************************************
* ns1 (Ns1Provider)
*   Update
*     <CnameRecord CNAME 3600, relative.exxampled.com., www.> ->
*     <CnameRecord CNAME 3600, relative.exxampled.com., www> (config)
*   Summary: Creates=0, Updates=1, Deletes=0, Existing Records=3
********************************************************************************
```

Anyway, going to back off on trying to get this into `1.0.0rc1` and cut that release w/o this change.

Not [super clear what the use cases for this is](https://github.com/octodns/octodns/issues/987#issuecomment-1641122007) and it looks like it would require changes to providers to support as well.

**Comment by istr** [comment]
@ross I am a little bit afraid that we would need to update lots of (or all) of the providers to support this.
You already said the same back in July:

> Not [super clear what the use case for this is](https://github.com/octodns/octodns/issues/987#issuecomment-1641122007) and it looks like it would require changes to providers to support as well.

I think that the primary use case for relative values is a shortcut/simplification for complicated master files.

When serving or submitting to providers through their API, the merging needs to be done  no matter what.

In our case, we would either need to implement the merging in the core. In this case, the few providers like `octodns-bind` with native support for relative entries will have to revert the merge. The other option would be to implement this in the vast majority of providers (since most of them do not support relative names through their API or even interpret relative values as absolute values, making things complicated/worse).

tl;dr - I would refrain from implementing this, as it could have nasty/breaking implications for many providers.

**Review by istr — COMMENTED** [review]
_If_ we implement this, I would change the check for "is relative" to be `endswith('.')` throughout the code. Maybe even tuck it away in a single function to be sure that there is a consistent check.

**Comment by ross** [comment]
Yeah. My plan for now at least, is to just let this sit here to see if the original discussion went anywhere or I was misunderstanding something. The assumption was that it'd likely just close out once the stale action decided it was time.
